Add extraction of comments from files.

tikerlade opened the pull request:
Add Tree-sitter nodes for all supported languages. Add `get_comments_from_file` function to TreeSitterParser for extracting comments from files.

tikerlade committed:
Add extraction of comments from files.

EgorBu commented on buckwheat/tokenizer.py:
```diff
@@ -249,6 +264,29 @@ def get_object_from_node(object_type: ObjectTypes, code: bytes, node: tree_sitte
+        """
+        Given a file and language of this file. Extract comments from this file.
+        :param file: the path to file.
+        :param lang: the language of code.
```
```suggestion
        :param lang: programming language.
```

EgorBu reviewed (changes requested):
Everything looks good - let's add test for each language and merge it

tikerlade committed:
Add comments extraction using Pygments and add tests.

Generalize get_comments_from_file function for all supported languages. Add comments to test files. Provide test data to test_tokenizer.

tikerlade committed:
Fixed typo in buckwheat/tokenizer.py

Co-authored-by: Egor Bulychev <egorbu@gmail.com>

tikerlade committed:
Fix another typo in buckwheat/tokenizer.py

tikerlade commented:
Edited PygmentsParser to extract comments and added tests.

EgorBu commented on buckwheat/tests/test_tokenizer.py:
```diff
@@ -5,95 +5,149 @@
                          {"out": 4, "tmp": 4, "string": 4, "get": 2, "block": 2, "new": 2,
-                          "stdout": 1, "stderr": 1}],
+                          "stdout": 1, "stderr": 1},
+                         ["# Declaration of get_stdout function", "=begin\nDeal with stderr\n=end"]],
```
Is it possible to extract comments without `/`, `#`, etc?

tikerlade replied:
I haven't find any solution for this task neither in tree-sitter nor in pygments. I think it may be done using postprocessing after extracting comments using `get_comments_from_file` method for different cases differently.

EgorBu commented on buckwheat/tests/test_files/test.py:
```diff
@@ -2,6 +2,7 @@
 
 from tree_sitter import Language, Parser
 
+# No parsers need to be declared here
```
do we have single and multi-line comments for each language?

EgorBu replied:
https://twitter.com/gvanrossum/status/112670605505077248

tikerlade replied:
Add multi-line comments for all languages except Shell. Out of the box Shell doesn't support multi-line comments.

EgorBu replied:
what is the type of this multiline string/comment in python?
```
"""
bla
"""
```

EgorBu commented on buckwheat/tests/test_tokenizer.py:
```diff
@@ -5,95 +5,149 @@
                           "socket": 1, "ipaddress": 1, "port": 1, "system": 1, "enumer": 1,
-                          "fatal": 1, "error": 1}],
+                          "fatal": 1, "error": 1},
+                         ["//", "/", " ", "D", "e", "f", "i", "n", "i", "n", "g", " ",
```
this comment looks strange - does it mean that we can extract comments in swift only as a list of chars?

tikerlade replied:
Remove Swift from comment extraction supported languages.

EgorBu commented on buckwheat/tests/test_tokenizer.py:
```diff
@@ -5,95 +5,149 @@
                           "not": 1, "donot": 1, "use": 1, "zend": 1, "alloc": 1, "range": 1,
-                          "printf": 1, "exists": 1}],
+                          "printf": 1, "exists": 1},
+                         ["// read file to contents variable",
```
let's create dataclass for this structure: language, filename, identifier counts, comments

tikerlade replied:
Great idea!

EgorBu reviewed (changes requested):
Great work! Almost done - I put several questions + let's make dataclass for a structure that contains all required information in a test

tikerlade committed:
Add classes for tests, add multi-lilne comments for Python.

Fix some minor typos. Add TestSample class for tests examples.
Add supporting of multi-line comments for Python.
Add more tests of multi-line and single-line comments for different languages.
Remove Swift from comment extraction supported languages.

tikerlade commented:
It seems like I have fixed your suggestions.

EgorBu commented on buckwheat/tests/test_files/test.cpp:
```diff
@@ -1,7 +1,7 @@
-  float mini = bs::Zero<float>();
-  float maxi = bs::Valmax<float>();
+  float mini = bs::Zero<float>();   // Initialize mini with zero
+  float maxi = bs::Valmax<float>(); /* Initialize maxi with maxval */
```
it would be good to split  this comment on 2 lines

EgorBu commented on buckwheat/tests/test_tokenizer.py:
```diff
@@ -3,97 +3,223 @@
 import unittest
 
-from ..tokenizer import get_identifiers_sequence_from_file
+from ..tokenizer import get_identifiers_sequence_from_file, get_comments_from_file
```
let's use absolut imports

EgorBu commented on buckwheat/tokenizer.py:
```diff
@@ -505,6 +569,23 @@ def get_classes_from_file(file: str, lang: str, identifiers_verbose: bool = Fals
+    """
+    lang = recognize_language_file(file)['language']
+
+    if lang in TreeSitterParser.COMMENTS.keys():
```
you can simplify it a bit 
```suggestion
    if lang in TreeSitterParser.COMMENTS:
```

EgorBu reviewed (changes requested):
Just couple more things - and it's done!)

tikerlade committed:
Minor fixes. Update test for python. Absolute paths instead of relative in test_tokenizer.

EgorBu commented on buckwheat/tokenizer.py:
```diff
@@ -333,13 +371,21 @@ class PygmentsParser:
+    COMMENTS = {"Scala": {pygments.token.Comment},
+                "Kotlin": {pygments.token.Comment},
+                "Haskell": {pygments.token.Comment},
+                "Python": {pygments.token.Comment, pygments.token.Literal.String.Doc}}
```
why do we use Pygment for python? Python is supported by tree-sitter

tikerlade replied:
because in tree-sitter we cannot extract multi-line comments and multi-line strings which represents comments cannot be extracted either

areyde replied:
If that's true, this showcases the advantages of having the second parser in the pipeline. TBF, comments do not really really require AST at all, so.

vsmaxim replied:
> because in tree-sitter we cannot extract multi-line comments and multi-line strings which represents comments cannot be extracted either

That's not tree sitter problem, python has no comments in it's [grammars](https://docs.python.org/3/reference/grammar.html), so they're not interpreted (or included in AST). 

But there is still a way to extract multiline-comments (which are just expressions with only multiline strings in it) with tree_sitter:
```
>>> multiline_string = '''
... """
... This is multiline comment
... Hmmm yeah
... """
... 
... a = [1, 2, 3]
... '''
>>> cur = parser.parse(multiline_string.encode()).walk()
>>> cur.goto_first_child()
True
>>> cur.node
<Node kind=expression_statement, start_point=(1, 0), end_point=(4, 3)>
>>> cur.goto_first_child()
True
>>> cur.node
<Node kind=string, start_point=(1, 0), end_point=(4, 3)>
```

That's obviously not good to extract only comments done with multiline strings and I think pygments should be used for comments extraction.

tikerlade committed:
Add lang parameter to get_comments_from_file

tikerlade committed:
Add usage of lang to get_comments_from_file tests in test_tokenizer.py

bzz commented:
Stumbled up on this functionality that seems useful for several use-cases (and seems to be actually used internally).

It would be interesting to know the resolution/reason why this was not merged.
